core.js

```diff
@@ -60,16 +44,92 @@ function _check(buffer, headers, options) {
 }
 
 export async function fileTypeFromTokenizer(tokenizer) {
-	try {
-		return new FileTypeParser().parse(tokenizer);
-	} catch (error) {
-		if (!(error instanceof strtok3.EndOfStreamError)) {
-			throw error;
-		}
+	await new FileTypeParser().fileTypeFromTokenizer(tokenizer);
+}
+
+export class TokenizerPositionError extends Error {
+	constructor(message = 'Tokenizer position changed during operation. This is not allowed as it makes following detectors read from the wrong offset. If the offset change is intentional, use recursion.') {
+		super(message);
+		this.name = 'TokenizerPositionError';
 	}
 }
 
-class FileTypeParser {
+export class FileTypeParser {
+	/**
+	 * The detectionImpossible below can be returned by custom detectors
+	 * when they are certain that the file type can not be determined,
+	 * neither by the detector itself, nor by any further detector.
+	 * This will interrupt the file type detection.
+	 */
+	static detectionImpossible = 'detection_impossible';
+
+	constructor(options) {
+		this.detectors = options?.customDetectors;
+
+		this.runCustomDetectors = this.runCustomDetectors.bind(this);
+		this.fileTypeFromTokenizer = this.fileTypeFromTokenizer.bind(this);
+		this.fileTypeFromBuffer = this.fileTypeFromBuffer.bind(this);
+		this.parse = this.parse.bind(this);
+	}
+
+	async runCustomDetectors(tokenizer) {
+		const {detectors} = this;
+		const initialPosition = tokenizer?.position;
+
+		for (const detector of detectors || []) {
+			const fileType = await detector(tokenizer);
+			if (fileType) {
+				return fileType;
+			}
+
+			if (initialPosition !== tokenizer?.position) {
+				throw new TokenizerPositionError();
+			}
+		}
+
+		return undefined;
+	}
+
+	async fileTypeFromTokenizer(tokenizer) {
```

Comment:
Member function of the class do not require the `fileType` prefix, which was introduced when we switched to ESM. As in the CJS the functions were typically used in combination with the module reference

```suggestion
	async fromTokenizer(tokenizer) {
```